src/transport/SessionManager.cpp

```diff
@@ -387,6 +387,17 @@ void SessionManager::ExpireAllPASEPairings()
     });
 }
 
+void SessionManager::ReleaseSessionForNodeExceptOne(const ScopedNodeId & node, const SessionHandle & stay)
```

Comment:
```suggestion
void SessionManager::ReleaseSessionsForNodeExceptOne(const ScopedNodeId & node, const SessionHandle & exception)
```
That said, we're going to still need to drop that session, just later, once we've sent the response, right?

And in particular, since `MarkForRemoval` does not actually remove the session until later when no one is holding refs to it, do we even need to except the one session involved?

Reply:
The problem is that calling `MarkForRemoval` will evict all weak-references, which are the type EC's hold today (which means the EC is going to terminated)...

Reply:
Perhaps the exception session can just start a special timer with timeout > aggregated MRP backoff.  `MarkForRemoval` is called right away for all sessions except the exception/pairing session (mandate one per Fabric).  `MarkForRemoval` is only called on the pairing session after the timeout.  Alternatively, we can leverage the FailSafe timer to cleanup the pairing session.

Reply:
The best I could come up with as a solution so far is to have `ExchangeContext` have an additional `Optional<SessionHandle>` member that gets pinned to the underlying secure session through an intentional call on some new method call `ExchangeContext::HoldoffSessionClosure` (or something).

This will be called by `UpdateNOC`.

Consequently, in `ExchangeContext::OnSessionReleased`, you return immediately from the function if we find that we have a pinned session. That is a signal to keep the exchange open and valid, and to not remove it from the retrans table.

You go ahead and mark all matching sessions (including the current one) for removal. This will ensure that the current EC does not get closed, and that the session also, will continue to have a non-zero ref count and won't get evicted.

Later, when the ACK has been received on the EC, the `ReliableMessageMgr` will promptly remove the `RetransTableEntry` from its table, which should also, result in the ref-count on the EC (not the session) going to 0. That will trigger the deleter, which will trigger `~ExchangeContext`, which will eventually, give up its session reference, which will finally, cause the SecureSession to get deleted.

Reply:
This is not super pretty, because it really mucks our EC code with behavior specific to UpdateNOC, but honestly, I couldn't think of something better....

Reply:
What I like about your proposal over my timer-based one is that the pairing session will get removed faster.  Are we sure though that the complete final transaction is indeed over a single exchange?  Your pinned-session-in-exchange method looks fine to me.